Source/WebCore/rendering/svg/SVGInlineTextBox.cpp

```diff
@@ -411,31 +411,19 @@ TextRun SVGInlineTextBox::constructTextRun(const RenderStyle& style, const SVGTe
 
 bool SVGInlineTextBox::mapStartEndPositionsIntoFragmentCoordinates(const SVGTextFragment& fragment, unsigned& startPosition, unsigned& endPosition) const
 {
-    if (startPosition >= endPosition)
-        return false;
+    unsigned startFragment = fragment.characterOffset - start();
+    unsigned endFragment = startFragment + fragment.length;
 
-    ASSERT(fragment.characterOffset >= start());
-    unsigned offset = fragment.characterOffset - start();
-    unsigned length = fragment.length;
+    // Find intersection between the intervals: [startFragment..endFragment) and [startPosition..endPosition)
+    startPosition = std::max(startFragment, startPosition);
+    endPosition = std::min(endFragment, endPosition);
 
-    if (startPosition >= offset + length || endPosition <= offset)
+    if (startPosition >= endPosition)
         return false;
 
-    if (startPosition < offset)
-        startPosition = 0;
-    else {
-        ASSERT(startPosition >= offset);
-        startPosition -= offset;
-    }
+    startPosition -= startFragment;
+    endPosition -= startFragment;
 
-    if (endPosition > offset + length)
-        endPosition = length;
-    else {
-        ASSERT(endPosition >= offset);
-        endPosition -= offset;
-    }
-
-    ASSERT_WITH_SECURITY_IMPLICATION(startPosition < endPosition);
```

Comment:
I think it is important to mention that there are there coordinates system for the SVG text:

1. Line coordinates
2. InlineTextBox coordinates
3. Fragment coordinates

`SVGInlineTextBox` and `SVGTextFragment` are both in the `Line coordinates.` The `startPosition` and `endPosition` are both in _InlineTextBox coordinates_. And this function maps them to the _Fragment coordinates_.